recipe-client-addon/lib/PreferenceManagement.jsm

```diff
@@ -0,0 +1,395 @@
+    log.debug(`PreferenceManagement.markLastSeen(${recipeName})`);
+
+    const store = await this.getStorage();
+    if (!(recipeName in store.data[namespace])) {
+      throw new Error(`Could not find a preference recipe named "${recipeName}"`);
+    }
+
+    store.data[namespace][recipeName].lastSeen = new Date().toJSON();
+    store.saveSoon();
+  },
+
+  /**
+   * Given a recipe name, returns the relevant stored data, if any.
+   *
+   * @param  {String} recipeName
+   * @resolves When the operation is completed succesfully
+   * @rejects {Error}
+   *   If there is no stored recipe with the given name.
+   */
+  async get(recipeName) {
+    log.debug(`PreferenceManagement.get(${recipeName})`);
+    const store = await this.getStorage();
+    if (!(recipeName in store.data[namespace])) {
+      throw new Error(`Could not find a preference recipe named "${recipeName}"`);
+    }
+
+    // Return a copy so mutating it doesn't affect the storage.
+    return Object.assign({}, store.data[namespace][recipeName]);
+  },
+
+  /**
+   * Determines if a recipe's preference type is valid, and checks that the new
+   * value type matches the existing preference type (if any value exists).
+   *
+   * @param  {Object} recipe Recipe object to analyze
+   * @throws {Error} If preference type is invalid, or doesn't match existing type.
+   */
+  validateRecipePrefTypes(recipe) {
+    const { preferenceName, preferenceType } = recipe;
+
+    const prevPrefType = Services.prefs.getPrefType(preferenceName);
+    const givenPrefType = PREFERENCE_TYPE_MAP[preferenceType];
+
+    if (!preferenceType || !givenPrefType) {
+      throw new Error(`Invalid preferenceType provided (given "${preferenceType}")`);
+    }
+
+    if (prevPrefType !== Services.prefs.PREF_INVALID && prevPrefType !== givenPrefType) {
+      throw new Error(
+        `Previous preference value is of type "${prevPrefType}", but was given "${givenPrefType}" (${preferenceType})`
+      );
+    }
+  },
+
+  /**
+   * Saves a recipe object into locally-namespaced storage.
+   *
+   * @param  {Object} recipe    Recipe data to save into storage.
+   * @returns {Promise}
+   * @resolves When the operation is completed succesfully
+   * @rejects Javascript exception.
+   */
+  async saveRecipeData(recipe) {
+    log.debug("PreferenceManagement.saveRecipeData()");
+
+    const store = await this.getStorage();
+    store.data[namespace][recipe.name] = recipe;
+    store.saveSoon();
+  },
+
+  /**
+   * Returns an array of stored recipe objects which are NOT expired.
+   *
+   * @returns {Promise}
+   * @resolves When the operation is completed succesfully
+   * @rejects Javascript exception.
+   */
+  async getActiveChanges() {
+    log.debug("PreferenceManagement.getActiveChanges()");
+    const allPrefs = await this.getStoredRecipes();
+    // Return copies so mutating them doesn't affect the storage.
+    return allPrefs.filter(e => !e.expired).map(e => Object.assign({}, e));
```

Comment:
Well, both are public methods, and so I think the expectation that `getStoredRecipes` offers you copies will need to remain reasonable in the future even if we refactor or something. We also have tests for both of these methods returning copies that don't affect the originals, so I think just making the assumption is fine - tests will save us if someone inadvertently starts violating that assumption. If you update/add comments accordingly, it seems unlikely we will break it, and not copying again does avoid doing extra work for no benefit. :-)